src/Features/Core/Portable/RemoveUnusedMembers/AbstractRemoveUnusedMembersDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,337 @@
+        }
+
+        private static ImmutableArray<DiagnosticDescriptor> CreateDescriptors(bool forceEnableRules)
+        {
+            // TODO: Enable these rules by default once we have designed the Tools|Option location and UI for such code quality rules.
+
+            // IDE0051: "Remove unused members" (Symbol is declared but never referenced)
+            var removeUnusedMembersTitle = new LocalizableResourceString(nameof(FeaturesResources.Remove_Unused_Private_Members), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            var removeUnusedMembersMessage = new LocalizableResourceString(nameof(FeaturesResources.Remove_Unused_Private_Members_Message), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            var removeUnusedMembersRule = CreateDescriptor(
+                IDEDiagnosticIds.RemoveUnusedMembersDiagnosticId, removeUnusedMembersTitle, removeUnusedMembersMessage, configurable: true, enabledByDefault: forceEnableRules);
+            var removeUnusedMembersRuleWithFadingRule = CreateUnnecessaryDescriptor(
+                IDEDiagnosticIds.RemoveUnusedMembersDiagnosticId, removeUnusedMembersTitle, removeUnusedMembersMessage, configurable: true, enabledByDefault: forceEnableRules);
+
+            // IDE0052: "Remove unread members" (Value is written and/or symbol is referenced, but the assigned value is never read)
+            var removeUnreadMembersTitle = new LocalizableResourceString(nameof(FeaturesResources.Remove_Unread_Private_Members), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            var removeUnreadMembersMessage = new LocalizableResourceString(nameof(FeaturesResources.Remove_Unread_Private_Members_Message), FeaturesResources.ResourceManager, typeof(FeaturesResources));
+            var removeUnreadMembersRule = CreateDescriptor(
+                IDEDiagnosticIds.RemoveUnreadMembersDiagnosticId, removeUnreadMembersTitle, removeUnreadMembersMessage, configurable: true, enabledByDefault: forceEnableRules);
+            var removeUnreadMembersRuleUnnecessaryWithFadingRule = CreateUnnecessaryDescriptor(
+                IDEDiagnosticIds.RemoveUnreadMembersDiagnosticId, removeUnreadMembersTitle, removeUnreadMembersMessage, configurable: true, enabledByDefault: forceEnableRules);
+
+            return ImmutableArray.Create(removeUnusedMembersRule, removeUnusedMembersRuleWithFadingRule,
+                    removeUnreadMembersRule, removeUnreadMembersRuleUnnecessaryWithFadingRule);
+        }
+
+        private DiagnosticDescriptor RemoveUnusedMemberRule => SupportedDiagnostics[1];
+        private DiagnosticDescriptor RemoveUnreadMemberRule => SupportedDiagnostics[3];
+
+        public override bool OpenFileOnly(Workspace workspace) => false;
+
+        // We need to analyze the whole document even for edits within a method body.
+        public override DiagnosticAnalyzerCategory GetAnalyzerCategory() => DiagnosticAnalyzerCategory.SemanticDocumentAnalysis;
```

Comment:
this is the 'what'.  In general though, comments are great for the "why".  For example, someone coming along would see this and go... "but why do we need to analyze the whole document if only a method is edited".  Giving an example/explanation woudl be good.